File: 13. Graph/12. Bridges/Bridges.java
import java.util.*;
import java.lang.*;
import java.io.*;
import java.math.*;

class Graph {

    int n;
    ArrayList<ArrayList<Integer>> graph;
    static int time = 0;

    Graph (int n) {
        this.n = n;
        this.graph = new ArrayList<>();
        for (int i=0; i<n; i++) {
            graph.add(new ArrayList<Integer>());
        }
    }

    void addEdge (int a, int b) {
        graph.get(a).add(b);
        graph.get(b).add(a);
    }

    void bridges () {
        int[] parent = new int[n];
        Arrays.fill(parent, -1);
        int[] disc = new int[n];
        Arrays.fill(disc, -1);
        int[] low = new int[n];
        for (int i=0; i<n; i++) {
            if (disc[i] == -1) {
                dfs(i, disc, low, parent);
            }
        }
    }


    void dfs (int src, int[] disc, int[] low, int[] parent) {
        disc[src] = low[src] = ++time;
        for (int child: graph.get(src)) {
            if (disc[child] == -1) {
                parent[child] = src;
                dfs(child, disc, low, parent);
                low[src] = Math.min(low[src], low[child]);
                if (low[child] > disc[src]) {
                    System.out.println(src+" "+child);
                }
            }
            else if (parent[src] != child) {
                low[src] = Math.min(low[src], disc[child]);
            }
        }
    }

}


class Bridges {
	
    public static void main(String[] args) throws IOException {
    	Graph graph = new Graph(5);
        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 2);
        graph.addEdge(0, 3);
        graph.addEdge(3, 4);
        graph.bridges();
    }
}

File: 13. Graph/12. Bridges/bridges.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll	long long
#define pb	push_back
#define mp	make_pair
#define fp(i, n)	for (int i=0; i<(int)n; i++)
#define fpr(i, a, b)	for (int i=(int)a; i<=(int)b; i++)

void fio()
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
}

int timer = 0;

void dfs (vector<int> *gr, int src, int par, int *disc, int *low)
{
	disc[src] = low[src] = ++timer;
	for (int adj : gr[src]) {
		if (adj == par) {continue;}
		if (disc[adj] != -1) {
			low[src] = min(low[src], disc[adj]);
		}
		else {
			dfs(gr, adj, src, disc, low);
			low[src] = min(low[src], low[adj]);
			if (low[adj] > disc[src]) {
				cout << src << " - " << adj << "\n";
			}
		}
	}
}

void bridges (vector<int> *gr, int n)
{
	int disc[n];
	memset(disc, -1, sizeof(disc));
	int low[n];
	fp(i, n) {
		if (disc[i] == -1) {
			dfs(gr, i, -1, disc, low);
		}
	}
}


int main()
{
	fio();

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);
#endif

	int n = 5;
	vector<int> gr[n];
	gr[0].pb(1);
	gr[1].pb(0);
	gr[0].pb(2);
	gr[2].pb(0);
	gr[2].pb(1);
	gr[1].pb(2);
	gr[0].pb(3);
	gr[3].pb(0);
	gr[3].pb(4);
	gr[4].pb(3);

	bridges(gr, n);

	return 0;
}
